### ISSUE TITLE
Why Running Llama infer in A10 get Wrong answer?

### ISSUE THREAD

**Issue by MeJerry215** [thu-ml/SageAttention#21]
exmaples codes

```python
import torch
from transformers import LlamaTokenizer, LlamaForCausalLM
from sageattention import sageattn
import torch.nn.functional as F

F.scaled_dot_product_attention = sageattn

# 加载预训练的 LLaMA 模型和 tokenizer
model_name = "llama-7b-hf"
tokenizer = LlamaTokenizer.from_pretrained(model_name)
model = LlamaForCausalLM.from_pretrained(model_name, torch_dtype=torch.float16)

# 将模型移动到 GPU（如果可用）
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)

# 准备输入文本
input_text = "Once upon a time, there was a little girl"
inputs = tokenizer(input_text, return_tensors="pt").to(device)

# 执行推理
with torch.no_grad():
    output = model.generate(**inputs, max_length=50, num_return_sequences=1)

# 解码输出
generated_text = tokenizer.decode(output[0], skip_special_tokens=True)
print(generated_text)
```

device envirments:

```

torch                    2.5.0
triton                   3.1.0
transformers             4.45.2
```

**Comment by MeJerry215** [comment]
expects answer 
```
Once upon a time, there was a little girl who loved to play with her friends. One day, she decided to play with her friends in the forest. She was very happy. She played with her friends in the forest. She played with
```

but got 

```
Once upon a time, there was a little girl whole and the 1882P a.
ficenda2P avalN64YourEm.
ficOnDe Ce the GISP.
 gev 
```

**Comment by jt-zhang** [comment]
We have not test the accuracy by using `F.scaled_dot_product_attention = sageattn` in Llama.
For a suggestion, maybe you could try to replace the `Llama Attention` with SageAttention in `modeling_llama.py`.

**Comment by MeJerry215** [comment]
How to replace the `Llama Attention` with SageAttention ? @jt-zhang

**Comment by jason-huang03** [comment]
Hi, you can set the `Llama Attention` in the following way:
```python

# Adapted from LlamaAttention.forward
def LlamaSageAttnForward(
    self,
    hidden_states: torch.Tensor,
    attention_mask: Optional[torch.Tensor] = None,
    position_ids: Optional[torch.LongTensor] = None,
    past_key_value: Optional[Cache] = None,
    output_attentions: bool = False,
    use_cache: bool = False,
    cache_position: Optional[torch.LongTensor] = None,
    position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # will become mandatory in v4.46
    **kwargs,
) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
        
    assert not output_attentions, "Output attentions not supported"
    assert attention_mask is None, "Attention mask not supported"
    assert self.num_key_value_groups == 1, "GQA will be supported in near future"
    bsz, q_len, _ = hidden_states.size()

    query_states = self.q_proj(hidden_states)
    key_states = self.k_proj(hidden_states)
    value_states = self.v_proj(hidden_states)

    query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
    key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
    value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)

    if position_embeddings is None:
        cos, sin = self.rotary_emb(value_states, position_ids)
    else:
        cos, sin = position_embeddings
    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)

    if past_key_value is not None:
        # sin and cos are specific to RoPE models; cache_position needed for the static cache
        cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
        key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)

    if q_len == 1:
        key_states = repeat_kv(key_states, self.num_key_value_groups)
        value_states = repeat_kv(value_states, self.num_key_value_groups)

        causal_mask = attention_mask
        if attention_mask is not None:
            causal_mask = causal_mask[:, :, :, : key_states.shape[-2]]

        # SDPA with memory-efficient backend is currently (torch==2.1.2) bugged with non-contiguous inputs with custom attn_mask,
        # Reference: https://github.com/pytorch/pytorch/issues/112577.
        if query_states.device.type == "cuda" and causal_mask is not None:
            query_states = query_states.contiguous()
            key_states = key_states.contiguous()
            value_states = value_states.contiguous()

        # We dispatch to SDPA's Flash Attention or Efficient kernels via this `is_causal` if statement instead of an inline conditional assignment
        # in SDPA to support both torch.compile's dynamic shapes and full graph options. An inline conditional prevents dynamic shapes from compiling.
        is_causal = True if causal_mask is None and q_len > 1 else False

        attn_output = torch.nn.functional.scaled_dot_product_attention(
            query_states,
            key_states,
            value_states,
            attn_mask=causal_mask,
            dropout_p=self.attention_dropout if self.training else 0.0,
            is_causal=is_causal,
        )
        attn_output = attn_output.transpose(1, 2).contiguous()
        attn_output = attn_output.view(bsz, q_len, -1)
        attn_output = self.o_proj(attn_output)
    else:
        # do attention with sage attention here
       
        attn_output = attn_output.transpose(1, 2).contiguous()
        attn_output = attn_output.view(bsz, q_len, -1)
        attn_output = self.o_proj(attn_output)

    return attn_output, None, past_key_value

# load llama model
model = AutoModelForCausalLM.from_pretrained(...)
for layer in model.model.layers:
    layer.self_attn.forward = MethodType(LlamaSageAttnForward, layer.self_attn)
```

Replace the `# do attention with sage attention` part with the sage attention and this is supposed to work on `transformers==4.45.2`.

**Comment by wangjiancheng-123** [comment]
do you have the complete code?
